Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp

```diff
@@ -284,4 +270,90 @@ String HTMLInputElement::value_sanitization_algorithm(String value) const
+
+    // The legacy-pre-activation behavior for input elements are these steps:
+
+    // 1. If this element's type attribute is in the Checkbox state, then set
+    // this element's checkedness to its opposite value (i.e. true if it is
+    // false, false if it is true) and set this element's indeterminate IDL
+    // attribute to false.
+    // FIXME: Set indeterminate to false when that exists.
+    if (type() == "checkbox") {
+        set_checked(!checked(), ChangeSource::User);
+    }
+
+    // 2. If this element's type attribute is in the Radio Button state, then
+    // get a reference to the element in this element's radio button group that
+    // has its checkedness set to true, if any, and then set this element's
+    // checkedness to true.
+    if (type() == "radio") {
+        String name = this->name();
+
+        document().for_each_in_inclusive_subtree_of_type<HTML::HTMLInputElement>([&](auto& element) {
+            if (element.checked() && element.name() == name) {
+                m_legacy_pre_activation_behavior_checked_element_in_group = element;
+                return IterationDecision::Break;
+            }
+            return IterationDecision::Continue;
+        });
+
+        set_checked_within_group();
+    }
+}
+
+void HTMLInputElement::legacy_cancelled_activation_behavior()
+{
+    // The legacy-canceled-activation behavior for input elements are these steps:
+
+    // 1. If the element's type attribute is in the Checkbox state, then set the
+    // element's checkedness and the element's indeterminate IDL attribute back
+    // to the values they had before the legacy-pre-activation behavior was run.
+    if (type() == "checkbox") {
+        set_checked(m_before_legacy_pre_activation_behavior_checked, ChangeSource::Programmatic);
+    }
+
+    // 2. If this element 's type attribute is in the Radio Button state, then
+    // if the element to which a reference was obtained in the
+    // legacy-pre-activation behavior, if any, is still in what is now this
+    // element' s radio button group, if it still has one, and if so, setting
+    // that element 's checkedness to true; or else, if there was no such
+    // element, or that element is no longer in this element' s radio button
+    // group, or if this element no longer has a radio button group, setting
+    // this element's checkedness to false.
+    if (type() == "radio") {
+        String name = this->name();
+        bool did_reselect_previous_element = false;
+        if (m_legacy_pre_activation_behavior_checked_element_in_group) {
+            auto& element_in_group = *m_legacy_pre_activation_behavior_checked_element_in_group;
+            if (name == element_in_group.name()) {
+                element_in_group.set_checked_within_group();
+                did_reselect_previous_element = true;
+            }
+
+            m_legacy_pre_activation_behavior_checked_element_in_group = nullptr;
```

Comment:
If legacy_cancelled_activation_behavior is not called (i.e. JS did not call preventDefault), we will still hold onto the HTMLInputElement that's being held by m_legacy_pre_activation_behavior_checked_element_in_group. Is there a way to avoid this?

Reply:
The only way to do that would be to have some sort of callback like `legacy_cancelled_activation_behavior_was_not_called` on the element which allows us to do that cleanup, since we don't know how long the JS might run for and what it might do (so `WeakPtr` is out of the question). Or we could put the storing of the input group element on `EventDispatcher` directly but that sounds messier.